File: failling-diamonds/src/diamonds.java
import java.awt.*;
import java.io.FileReader;
import java.io.FileWriter;
import java.util.Map;
import java.util.Scanner;
import java.util.TreeMap;

/**
 * Created by MB on 7/26/2014.
 */
public class diamonds {


    static public String SAMPLE = "failling-diamonds/src/sample.in";
    static public String SMALL = "failling-diamonds/src/B-small-practice.in";
    static public String OUT = "failling-diamonds/src/output";

    static public int N ;
    static public int position[];



    static public int X;
    static public int Y;
    static public int recurse=0;

    static int test =0;

    public static void main(String[] args) throws Exception{


        Scanner in = new Scanner(new FileReader(SAMPLE));
        FileWriter out = new FileWriter(OUT);

        int nb_cases = in.nextInt();

        for(int i_case=0; i_case < nb_cases ; i_case++){

            N = in.nextInt();
            position = new int[2];
            position[0] = in.nextInt();
            position[1] = in.nextInt();
            int[][] field = new int[1500][1500];
            X = 1500/2 +1;
            Y = 0;
            position[0] = position[0] + X;


            for(int i=0; i< field.length;i++)
                for(int j=0; j<field.length; j++)
                    field[i][j] = 0;

            test=0;
            recurse = 0;
            int chances = find_chances(0,field);
            float a = test;
            float b= recurse;
            float per = a/b;
           // System.out.println("Case #"+(i_case+1)+": hits: " +chances + "   nbRecurs: "+recurse + "                    position :"+position[0]+" - "+position[1] +"  test:"+test +"res:"+per);
            System.out.println("Case #"+(i_case+1)+": "+per);
            out.write("Case #"+(i_case+1)+": "+per+"\n");
        }


    }



    public static int find_chances(int nb_diamonds, int[][] field){


        if(nb_diamonds == N){
            if (field[position[0]][position[1]]==1){
                test=test+1;
            }

            recurse +=1;
            return 0;
        }

        int total_hit=0;
        if( nb_diamonds == 0){
            field[X][Y] = 1;

            if(field[position[0]][position[1]] == 1) {
                total_hit += find_chances(nb_diamonds+1,field) +1;
            }else{
                total_hit += find_chances(nb_diamonds+1,field);
            }
        }else{

            int y = get_first_impactY(field);
            int[][] cpy_left = copyArray(field);
            if(slide_to(y,-1,cpy_left)){
                if(cpy_left[position[0]][position[1]] == 1){
                    total_hit += find_chances(nb_diamonds+1,cpy_left)+ 1;
                }else {
                    int[][] cpy_left2 = copyArray(cpy_left);
                    total_hit += find_chances(nb_diamonds+1,cpy_left2);
                }
            }
            int[][] cpy_right = copyArray(field);
            if(slide_to(y,1, cpy_right)){
                if(cpy_right[position[0]][position[1]] == 1){
                    total_hit += find_chances(nb_diamonds+1,cpy_right) +1;

                }else {
                    int[][] cpy_right2 = copyArray(cpy_right);
                    total_hit += find_chances(nb_diamonds+1,cpy_right2);
                }
            }
        }

        return total_hit;
    }

    public static boolean slide_to(int y1, int direction, int[][] field){
        boolean can_slide = false;

        int x=X;
        int y = y1;

        if (field[x+direction][y-1] != 1){
            while (true){
                if( y == 0 || (field[x-1][y-1] == 1 && field[x+1][y-1] == 1)){
                    field[x][y] = 1;
                    return true;
                }
                x += direction;
                y = y-1;
            }
        }

        return false;
        //return can_slide;
    }

    public static int get_first_impactY(int[][] field){
        int y=0;
        while(y < N){
            if (field[X][y] != 1){
                return y+1;
            }
            y++;
        }
        return 1;
    }
    public static int[][] copyArray(int[][] old){
        int[][] current= new int[old.length][old.length];
        for(int i=0; i<old.length; i++)
            for(int j=0; j<old[i].length; j++)
                current[i][j]=old[i][j];

        return current;
    }

    public class Key {

        private final int x;
        private final int y;

        public Key(int x, int y) {
            this.x = x;
            this.y = y;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (!(o instanceof Key)) return false;
            Key key = (Key) o;
            return x == key.x && y == key.y;
        }

        @Override
        public int hashCode() {
            int result = x;
            result = 31 * result + y;
            return result;
        }

    }

}

File: diamond-failling-scala/src/main/scala/diamonds.scala
import java.io.{FileReader, FileWriter}
import java.util.Scanner

/**
 * Created by MB on 7/27/2014.
 */

object Test {

  //val SAMPLE ="diamond-failling-scala/src/main/scala/sample.in"
  val SAMPLE ="src/main/scala/sample.in"
  val SMALL ="src/main/scala/B-small-practice.in"
  val LARGE ="src/main/scala/B-large-practice.in"
  val OUT ="src/main/scala/output"

  var N:Int = _
  var X:Int = _
  var Y:Int = _

  var hits:Int = _
  var total:Int =_


  def main(args: Array[String]) {
    // for loop execution with a range

    val in = new Scanner(new FileReader(LARGE))
    val out = new FileWriter(OUT)

    val nb_cases = in.nextInt()

    for(i <-0 until nb_cases by 1) {

      N = in.nextInt()
      X = in.nextInt()
      Y = in.nextInt()

      hits = 0
      total = 0

      var field = collection.mutable.Map[(Int, Int), Int]()
      var result = 0.toFloat;

      var t = getSommet()

      val abs_pos = Math.abs(X) + Math.abs(Y)
      if(abs_pos < t){
        result = 1.toFloat
      }else if (abs_pos > t){
        result = 0.toFloat
      }else{
        remplir(field,t-2)
        val n = N - field.size
        solve(0,n, field)
        result = hits.toFloat/total.toFloat
      }

      println("Case #"+(i+1)+": "+result)
     // println("hits :"+hits+" total :"+total + " test:" + test)
      out.write("Case #"+(i+1)+": "+result+"\n")
    }
    in.close()
    out.close()

  }
  def remplir(field:collection.mutable.Map[(Int,Int),Int],subPiramid:Int){

      field.put((0,0),1)
    if(!(subPiramid<=0)){
      while(!field.contains(0,subPiramid)){
        var left =  slide_to(-1,field)
        var right = slide_to(1,field)
        if( !left && !right){
          val y = getYimpact(field)
          field.put((0,y),1)
        }
      }

    }

    /* var y =0
     for (x <- subPiramid until 0 by -1){
         field.put((x,y),1)
       y+=1
     }
     y = subPiramid
     for (x <- 0 until -subPiramid by -1){
       field.put((x,y),1)
       y-=1
     }
 */
  }


  def getSommet(): Int = {
    var t=0;
    var n = 0

    while(true){
      n += 2*t+1
      if(n >= N){
        return t
      }
      t+=2
    }

    return t
  }

  def solve2(nb_to_place:Int, field:collection.mutable.Map[(Int,Int),Int]){
    for(i <- 0 until nb_to_place -1 by 1){
      slide_to(-1,field)

    }



  }

  def solve(i_diamond :Int, N:Int, field:collection.mutable.Map[(Int,Int),Int]) {

    if( i_diamond == N ) {
      if( field.contains(X,Y)) {
        hits += 1
      }
      total +=1
      return true;
    }
    if(false ){
      field.put((0,0),1)
      solve(i_diamond+1,N, field)
    }else{
      val field_left = collection.mutable.Map() ++= field
      val place_left:Boolean = slide_to(-1,field_left)
      if( (place_left)){
          solve(i_diamond+1,N,field_left)
      }

      val field_right = collection.mutable.Map() ++= field
      val place_right = slide_to(1,field_right)
      if(place_right){
        solve(i_diamond+1,N,field_right)
      }

      if(!place_left && !place_right){
        val y = getYimpact(field)
        field.put((0,y),1)
        solve(i_diamond+1,N,field)
      }

    }
  }


  def slide_to(direction:Int, field:collection.mutable.Map[(Int,Int),Int]): Boolean = {

      var x = 0
      var y = getYimpact(field)
      if(!field.contains(x+direction,y-1)){
        while(true){
          if( y==0 || (field.contains((x-1,y-1)) && field.contains((x+1,y-1)))  ){
            field.put((x,y),1)
            return true
          }
          x = x+ direction
          y += -1
        }
      }
      return false
  }


  def getYimpact(field: collection.mutable.Map[(Int,Int),Int]): Int = {
    var y = 0
    while(y < N ){
      val k = (0,y)
      if( !field.contains(k)){
        return y
      }
      y+=2
    }
    return y
  }



}


